File: 2 курс/Алгоритмы и структуры данных/tasksCpp/B_2.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<int> countLettersFrequency(const vector<char>& lettersList) {
  vector<int> lettersFrequency(26, 0);
  for (char ch : lettersList) {
    int x = tolower(ch);
    if (x >= 'a' && x <= 'z') {
      lettersFrequency[x - 'a']++;
    }
  }
  return lettersFrequency;
}

bool zooIsValid(const string& lettersInput, const vector<char>& lettersList) {
  if (lettersInput.length() % 2 != 0) {
    return false;
  }

  vector<int> lettersFrequency = countLettersFrequency(lettersList);
  for (size_t i = 0; i < lettersFrequency.size(); i++) {
    if (lettersFrequency[i] != 2 && lettersFrequency[i] != 0) {
      return false;
    }
  }

  return true;
}

void sortZoo(const string& lettersInput) {
  vector<char> lettersList;
  lettersList.reserve(lettersInput.length());
  for (size_t i = 0; i < lettersInput.length(); i++) {
    lettersList.push_back(lettersInput[i]);
  }

  if (!zooIsValid(lettersInput, lettersList)) {
    cout << "Impossible";
  } else {
    string answerStr;
    while (!lettersList.empty()) {
      size_t mid = lettersList.size() / 2 - 1;
      char cage = lettersList[mid];
      char animal = lettersList[mid + 1];
      for (size_t _ = 0; _ < lettersList.size(); ++_) {
        cage = lettersList[mid];
        animal = lettersList[mid + 1];
        if (cage != animal && tolower(cage) != tolower(animal)) {
          char first = lettersList.front();
          lettersList.erase(lettersList.begin());
          lettersList.push_back(first);
        } else {
          break;
        }
      }
      if (cage != animal && tolower(cage) == tolower(animal)) {
        answerStr += to_string(tolower(lettersList[mid]) - 'a' + 1) + " ";
        lettersList.erase(lettersList.begin() + static_cast<int>(mid) + 1);
        lettersList.erase(lettersList.begin() + static_cast<int>(mid));
      } else {
        cout << "Impossible";
        return;
      }
    }
    cout << "Possible\n";
    cout << answerStr;
  }
}

int main() {
  string lettersInput;
  cin >> lettersInput;
  sortZoo(lettersInput);

  return 0;
}

File: 2 курс/Алгоритмы и структуры данных/tasksPython/B_2.py
def countLettersFrequency(zooList):
    lettersFrequency = [0] * 26
    for i in range(len(zooList)):
        ind = ord(zooList[i].lower()) - ord("a")
        lettersFrequency[ind] += 1
    return lettersFrequency


def zooIsValid(lettersInput, lettersList):
    if len(lettersInput) % 2 != 0:
        return False

    lettersFrequency = countLettersFrequency(lettersList)
    for i in range(len(lettersFrequency)):
        if lettersFrequency[i] != 2 and lettersFrequency[i] != 0:
            return False
    return True


def sortZoo(lettersInput):
    lettersList = []
    for letter in lettersInput:
        lettersList.append(letter)
    if (lettersInput == "" or not zooIsValid(lettersInput, lettersList)):
        print("Impossible")
        return
    else:
        i = 0
        answerStr = ""
        while lettersList:
            i += 1
            mid = len(lettersList) // 2 - 1
            cage = lettersList[mid]
            animal = lettersList[mid + 1]
            for _ in range(len(lettersList)):
                if (cage != animal and cage.lower() != animal.lower()):
                    lettersList = lettersList[1:] + lettersList[:1]
                    cage = lettersList[mid]
                    animal = lettersList[mid + 1]
                else:
                    break

            if (cage != animal and cage.lower() == animal.lower()):
                answerStr += str(ord(lettersList[mid].lower()) - ord('a') + 1) + " "
                lettersList.remove(cage)
                lettersList.remove(animal)
            else:
                print("Impossible")
                return
        print("Possible\n" + answerStr)


if __name__ == "__main__":
    lettersInput = input()
    sortZoo(lettersInput)
